README.md
# Atomic.js

Atomic is a HTML5 circular progress bar. It functions with
minimal use of DOM elements, CSS and JS. Atomic will be used by the
getting started pages of [Nodejitsu](https://www.nodejitsu.com/) and
[observe.it](https://observe.it).

### Installation

To include atomic.js in your project, include it in your [browserify] build.
For standalone usage gulp builds are available to create minified and
obfuscated output.

```bash
npm run-script compile       # create minified and smallest assets
npm run-script compile-dev   # to compile dev dependencies with sourcemaps included
```

[browserify]: http://browserify.org/

### How to use

Atomic works by combining CSS sibling power with the animation capacties of the
HTML5 `canvas` element. To use Atomic include the html below and make sure to
add the CSS and JS assets. The example below will generate a progress bar with 5
steps. Removing or adding list items will adjust the progress bar accordingly.

![Atomic](https://raw.github.com/Swaagie/atomic/master/demo/atomic.png)

I recommend to use the [Montserrat font][font] of Google Webfonts, as the styling
is optimized for it and it just looks more pretty. See the demo page for the
actual inclusion.

The label text is displayed on `hover` and replaces the number of the step. This
could be used to display a title for instance.

```html
  <section class="atomic">
    <input type="text" value="54" size="3" maxlength="3">
    <ol>
      <li><label>Step 1</label></li>
      <li><label>More descriptive</label></li>
      <li><label>labels are fun</label></li>
      <li><label>Nice steps</label></li>
      <li><label>Almost there</label></li>
      <li><label>The end</label></li>
    </ol>
    <canvas></canvas>
  </section>
```

The required CSS and JS assets are small.
 - **atomic.min.js** compressed: 2.5kB minified, 1.1kB gzipped
 - **atomic.min.css** compressed: 6.1kB minified, 1.1kB gzipped.

[font]: http://www.google.com/fonts#UsePlace:use/Collection:Montserrat

### Live Demo

Changing the value of the input will animate the progress inidicator and select
the proper step alongside. A demonstration can be found on the
[demo page](http://www.martijnswaagman.nl/atomic/).

### Browser support

Latest chrome, safari, firefox and IE9+

### Tests

TODO


CHANGELOG.md
## Changelog

### 2.0.0
- Add browserify build
- Use gulp to build the files in stead of Square
- Defer initialization to init script

### 1.4.0
- Calculate circular positions instead of transforming

### 1.3.2
- Fix label content positioning on hover if unchecked
- Allow radio's to exist outside atomic with external flag

### 1.3.1
- Unchecked first step if the progress is 0

### 1.3.0
- Move defaults to external file for easy customization

### 1.2.0
- Increase maximum number of steps to 9
- Calculate electron/step size by 2 * golden ratio

### 1.1.3
- fix relative positioning for firefox
- checked functioning against firefox, opera, chrome, chromium, safari

### 1.1.2
- fix rendering artifacts due to rounding errors

### 1.1.1
- round the numbers displayed in the input
- proper checked radio button on initialization

### 1.1.0
- animated electrons/steps
- labels show text on hover
- better centered step numbers


index.js
'use strict';

/**
 * Polyfill for cancel & requestAnimationFrame for smooth and fast animations.
 *
 * @api private
 */
(function polyfill() {
  var expectedTime = 0
    , vendors = ['ms', 'moz', 'webkit', 'o']
    , vendor;

  while(!window.requestAnimationFrame && (vendor = vendors.pop())) {
    window.requestAnimationFrame = window[vendor + 'RequestAnimationFrame'];
    window.cancelAnimationFrame = window[vendor + 'CancelAnimationFrame']
     || window[vendor+'CancelRequestAnimationFrame'];
  }

  if (!window.cancelAnimationFrame) window.cancelAnimationFrame = clearTimeout;
  if (!window.requestAnimationFrame) {
    window.requestAnimationFrame = function requestAnimationFrame(callback) {
      var currentTime = Date.now()
        , adjustedDelay = 16 - (currentTime - expectedTime)
        , delay = adjustedDelay > 0 ? adjustedDelay : 0;

      expectedTime = currentTime + delay;
      return setTimeout(function fps() {
        callback(expectedTime);
      }, delay);
    };
  }
})();

/**
 * Constructor of Atomic, called only if the progress bar is in the page.
 *
 * @Constructor
 * @param {Element} atom
 * @param {Element} parent attach progress radio's to specific element
 * @api public
 */
function Atomic(atom, parent) {
  var canvas = atom.getElementsByTagName('canvas')[0]
    , size = getComputedStyle(canvas).getPropertyValue('width')
    , π = Math.PI;

  // Set some properties for the canvas to work with.
  this.ctx = canvas.getContext('2d');
  this.color = getComputedStyle(atom).getPropertyValue('color');
  this.text = atom.getElementsByTagName('input')[0];
  this.value = 0;
  this.parent = parent || atom;

  // Some defaults required for the animations
  this.r = +size.replace('px', '') / 2;
  this.ψ = π / 50;

  // Set height and width on the canvas.
  canvas.setAttribute('width', size);
  canvas.setAttribute('height', size);

  // Initialize and start the animation cycle for the first time.
  this.initialize(atom).animate(this.value, +this.text.value);
}

/**
 * Initialize event listeners and add input elements.
 *
 * @param {Element} atom
 * @returns {Atomic} fluent interface
 * @api private
 */
Atomic.prototype.initialize = function initialize(atom) {
  var electrons = atom.getElementsByTagName('li')
    , attributes = { type: 'radio', name: 'atomic' }
    , element, step, checked, n, i;

  // Attach onChange listeners to the text element
  this.text.addEventListener('change', this.listen.bind(this));

  // Step size and radio buttons.
  this.i = i = electrons.length;
  this.n = n = 100 / i;
  this.radio = [];

  // Set data-n on section to #list items.
  atom.setAttribute('data-n', i);

  // Add radiobuttons and listen to any clicks on the labels
  while (i--) {
    this.radio.unshift(element = document.createElement('input'));
    element.addEventListener('click', this.listen.bind(this));

    step = n * i;
    checked = step - this.text.value;

    attributes.id = 'i' + i;
    attributes.value = step;
    element.checked = checked > 0 && checked <= n && +this.text.value;

    this.parent.insertBefore(this.setAttributes(element, attributes), this.parent.firstChild);
    electrons[i].getElementsByTagName('label')[0].setAttribute('for', 'i' + i);
  }

  return this;
};

/**
 * Small helper function to set attributes on created elements.
 *
 * @param {Element} element
 * @param {Object} attributes
 * @returns {Element}
 * @api private
 */
Atomic.prototype.setAttributes = function setAttributes(element, attributes) {
  for(var key in attributes) {
    element.setAttribute(key, attributes[key]);
  }

  return element;
};

/**
 * Initialize animate function with specified start/end.
 *
 * @param {Number} start
 * @param {Number} end
 * @returns {Element}
 * @api private
 */
Atomic.prototype.animate = function animate(start, end) {
  var self = this
    , cw = end > start ? 1 : -1
    , π = Math.PI
    , Δ = this.ψ * cw
    , θ = Δ / 2
    , α = π / -2 + this.ψ * start
    , β = π / -2 + this.ψ * end
    , steps = Math.abs(Math.ceil((β - α) / Δ))
    , ctx = this.ctx
    , r = this.r
    , λ = 0;

  /**
   * Render each slice.
   *
   * @api private
   */
  function render() {
    var corr = (start && start !== 100) || λ ? θ : 0
      , r1 = α + λ * Δ - corr
      , r2 = r1 + corr + Δ;

    // Limit animation, final step might be one to much.
    if (~cw && r2 > β) r2 = β;
    if (!~cw && r2 < β) r2 = β;

    // Start drawing the circle slice.
    ctx.beginPath();
    ctx.globalCompositeOperation = 'destination-' + (~cw ? 'over' : 'out');
    ctx.fillStyle = ~cw ? self.color : 'rgba(0, 0, 0, 1)';
    ctx.arc(r, r, r, r1, r2, !~cw);
    ctx.lineTo(r, r);
    ctx.closePath();
    ctx.fill();

    // Update the current value to reflect the running endpoint of the animation.
    self.value = start + 1 * cw + λ++ * cw;

    // stop calling render as soon as steps are done or if equal to end.
    if (λ < steps && r2 !== β) return self.id = window.requestAnimationFrame(render);

    // If complet animation is rendered make sure the value is the provided end.
    // This will fix some rendering artifacts due to rounding errors.
    self.value = end;
  }

  this.id = window.requestAnimationFrame(render);
};

/**
 * Change the current value of the progress bar and selected radio buttons.
 *
 * @param {Number} end
 * @returns {Atomic} fluent interface
 * @api public
 */
Atomic.prototype.update = function update(end) {
  if (end === this.value) return this;
  if (end > 100) end = 100;

  // Which radio button should be selected.
  var current = Math.floor(end / this.n)
    , max = this.i - 1;

  // Update the input value and selected radio button.
  end = Math.round(end);
  this.text.value = end;
  this.radio[current < max ? current : max].checked = true;

  // Stop running animations and start fresh one.
  window.cancelAnimationFrame(this.id);
  this.animate(this.value, end);

  return this;
};

/**
 * Event listener for any changes on the inputs.
 *
 * @param {Event} e
 * @returns {Atomic} fluent interface
 * @api private
 */
Atomic.prototype.listen = function listen(e) {
  this.update(+(e.target || e.srcElement).value);

  return this;
};

/**
 * Initialize each progress bar on the page.
 *
 * @api public
 */
Atomic.init = function init() {
  var atoms = document.querySelectorAll('[data-atomic]');

  for (var k = 0; k < atoms.length; k++) {
    Atomic[k] = new Atomic(atoms[k]);
  }
};

//
// Expose module.
//
module.exports = Atomic;

gulpfile.js
'use strict';

var nib = require('nib')
  , gulp = require('gulp')
  , concat = require('gulp-concat')
  , uglify = require('gulp-uglify')
  , stylus = require('gulp-stylus')
  , minify = require('gulp-minify-css')
  , sourcemaps = require('gulp-sourcemaps');

//
// Set production flag, useful to exclude sourcemaps.
//
var production = process.env.NODE_ENV === 'production';

//
// Default set of tasks.
//
gulp.task('default', ['script', 'style']);

//
// Define Gulp task to concat and minifiy scripts.
//
gulp.task('script', function script() {
  var work = gulp.src('index.js')
    .pipe(sourcemaps.init())
    .pipe(uglify())
    .pipe(concat(type('atomic.{type}.js')));

  //
  // Only write the source map if we are not building for production.
  //
  if (!production) work.pipe(sourcemaps.write());
  return work.pipe(gulp.dest('dist'));
});

//
// Gulp task to generate CSS from stylus files.
//
gulp.task('style', function style() {
  var work = gulp.src('css/*.styl')
    .pipe(stylus({ use: nib() }))
    .pipe(concat(type('atomic.{type}.css')));

  //
  // Only minify if env is production.
  //
  if (production) work.pipe(minify());
  return work.pipe(gulp.dest('dist'));
});

function type(string) {
  return string.replace('{type}', production ? 'min' : 'dev');
}

src/atomic.js
(function initAtomic(w, d) {
  'use strict';

  /**
   * Polyfill for cancel & requestAnimationFrame for smooth and fast animations.
   *
   * @api private
   */
  (function polyfill() {
    var expectedTime = 0
      , vendors = ['ms', 'moz', 'webkit', 'o']
      , vendor;

    while(!w.requestAnimationFrame && (vendor = vendors.pop())) {
      w.requestAnimationFrame = w[vendor + 'RequestAnimationFrame'];
      w.cancelAnimationFrame = w[vendor + 'CancelAnimationFrame']
       || w[vendor+'CancelRequestAnimationFrame'];
    }

    if (!w.cancelAnimationFrame) w.cancelAnimationFrame = clearTimeout;
    if (!w.requestAnimationFrame) {
      w.requestAnimationFrame = function requestAnimationFrame(callback) {
        var currentTime = Date.now()
          , adjustedDelay = 16 - (currentTime - expectedTime)
          , delay = adjustedDelay > 0 ? adjustedDelay : 0;

        expectedTime = currentTime + delay;
        return setTimeout(function fps() {
          callback(expectedTime);
        }, delay);
      };
    }
  })();

  /**
   * Constructor of Atomic, called only if the progress bar is in the page.
   *
   * @Constructor
   * @param {Element} atom
   * @param {Element} parent attach progress radio's to specific element
   * @api public
   */
  function Atomic(atom, parent) {
    var canvas = atom.getElementsByTagName('canvas')[0]
      , size = getComputedStyle(canvas).getPropertyValue('width')
      , π = Math.PI;

    // Set some properties for the canvas to work with.
    this.ctx = canvas.getContext('2d');
    this.color = getComputedStyle(atom).getPropertyValue('color');
    this.text = atom.getElementsByTagName('input')[0];
    this.value = 0;
    this.parent = parent || atom;

    // Some defaults required for the animations
    this.r = +size.replace('px', '') / 2;
    this.ψ = π / 50;

    // Set height and width on the canvas.
    canvas.setAttribute('width', size);
    canvas.setAttribute('height', size);

    // Initialize and start the animation cycle for the first time.
    this.initialize(atom).animate(this.value, +this.text.value);
  }

  /**
   * Initialize event listeners and add input elements.
   *
   * @param {Element} atom
   * @returns {Atomic} fluent interface
   * @api private
   */
  Atomic.prototype.initialize = function initialize(atom) {
    var electrons = atom.getElementsByTagName('li')
      , attributes = { type: 'radio', name: 'atomic' }
      , element, step, checked, n, i;

    // Attach onChange listeners to the text element
    this.text.addEventListener('change', this.listen.bind(this));

    // Step size and radio buttons.
    this.i = i = electrons.length;
    this.n = n = 100 / i;
    this.radio = [];

    // Set data-n on section to #list items.
    atom.setAttribute('data-n', i);

    // Add radiobuttons and listen to any clicks on the labels
    while (i--) {
      this.radio.unshift(element = d.createElement('input'));
      element.addEventListener('click', this.listen.bind(this));

      step = n * i;
      checked = step - this.text.value;

      attributes.id = 'i' + i;
      attributes.value = step;
      element.checked = checked > 0 && checked <= n && +this.text.value;

      this.parent.insertBefore(this.setAttributes(element, attributes), this.parent.firstChild);
      electrons[i].getElementsByTagName('label')[0].setAttribute('for', 'i' + i);
    }

    return this;
  };

  /**
   * Small helper function to set attributes on created elements.
   *
   * @param {Element} element
   * @param {Object} attributes
   * @returns {Element}
   * @api private
   */
  Atomic.prototype.setAttributes = function setAttributes(element, attributes) {
    for(var key in attributes) {
      element.setAttribute(key, attributes[key]);
    }

    return element;
  };

  /**
   * Initialize animate function with specified start/end.
   *
   * @param {Number} start
   * @param {Number} end
   * @returns {Element}
   * @api private
   */
  Atomic.prototype.animate = function animate(start, end) {
    var self = this
      , cw = end > start ? 1 : -1
      , π = Math.PI
      , Δ = this.ψ * cw
      , θ = Δ / 2
      , α = π / -2 + this.ψ * start
      , β = π / -2 + this.ψ * end
      , steps = Math.abs(Math.ceil((β - α) / Δ))
      , ctx = this.ctx
      , r = this.r
      , λ = 0;

    /**
     * Render each slice.
     *
     * @api private
     */
    function render() {
      var corr = (start && start !== 100) || λ ? θ : 0
        , r1 = α + λ * Δ - corr
        , r2 = r1 + corr + Δ;

      // Limit animation, final step might be one to much.
      if (~cw && r2 > β) r2 = β;
      if (!~cw && r2 < β) r2 = β;

      // Start drawing the circle slice.
      ctx.beginPath();
      ctx.globalCompositeOperation = 'destination-' + (~cw ? 'over' : 'out');
      ctx.fillStyle = ~cw ? self.color : 'rgba(0, 0, 0, 1)';
      ctx.arc(r, r, r, r1, r2, !~cw);
      ctx.lineTo(r, r);
      ctx.closePath();
      ctx.fill();

      // Update the current value to reflect the running endpoint of the animation.
      self.value = start + 1 * cw + λ++ * cw;

      // stop calling render as soon as steps are done or if equal to end.
      if (λ < steps && r2 !== β) return self.id = w.requestAnimationFrame(render);

      // If complet animation is rendered make sure the value is the provided end.
      // This will fix some rendering artifacts due to rounding errors.
      self.value = end;
    }

    this.id = w.requestAnimationFrame(render);
  };

  /**
   * Change the current value of the progress bar and selected radio buttons.
   *
   * @param {Number} end
   * @returns {Atomic} fluent interface
   * @api public
   */
  Atomic.prototype.update = function update(end) {
    if (end === this.value) return this;
    if (end > 100) end = 100;

    // Which radio button should be selected.
    var current = Math.floor(end / this.n)
      , max = this.i - 1;

    // Update the input value and selected radio button.
    end = Math.round(end);
    this.text.value = end;
    this.radio[current < max ? current : max].checked = true;

    // Stop running animations and start fresh one.
    w.cancelAnimationFrame(this.id);
    this.animate(this.value, end);

    return this;
  };

  /**
   * Event listener for any changes on the inputs.
   *
   * @param {Event} e
   * @returns {Atomic} fluent interface
   * @api private
   */
  Atomic.prototype.listen = function listen(e) {
    this.update(+(e.target || e.srcElement).value);

    return this;
  };

  // Initialize each progress bar on the page.
  var atoms = d.getElementsByClassName('atomic');
  for (var k = 0; k < atoms.length; k++) {
    Atomic[k] = new Atomic(atoms[k]);
  }

  // Expose constructor and current instances to window.
  w.Atomic = Atomic;
})(window, document);


demo/bundle.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
require('../').init();
},{"../":2}],2:[function(require,module,exports){
'use strict';

/**
 * Polyfill for cancel & requestAnimationFrame for smooth and fast animations.
 *
 * @api private
 */
(function polyfill() {
  var expectedTime = 0
    , vendors = ['ms', 'moz', 'webkit', 'o']
    , vendor;

  while(!window.requestAnimationFrame && (vendor = vendors.pop())) {
    window.requestAnimationFrame = window[vendor + 'RequestAnimationFrame'];
    window.cancelAnimationFrame = window[vendor + 'CancelAnimationFrame']
     || window[vendor+'CancelRequestAnimationFrame'];
  }

  if (!window.cancelAnimationFrame) window.cancelAnimationFrame = clearTimeout;
  if (!window.requestAnimationFrame) {
    window.requestAnimationFrame = function requestAnimationFrame(callback) {
      var currentTime = Date.now()
        , adjustedDelay = 16 - (currentTime - expectedTime)
        , delay = adjustedDelay > 0 ? adjustedDelay : 0;

      expectedTime = currentTime + delay;
      return setTimeout(function fps() {
        callback(expectedTime);
      }, delay);
    };
  }
})();

/**
 * Constructor of Atomic, called only if the progress bar is in the page.
 *
 * @Constructor
 * @param {Element} atom
 * @param {Element} parent attach progress radio's to specific element
 * @api public
 */
function Atomic(atom, parent) {
  var canvas = atom.getElementsByTagName('canvas')[0]
    , size = getComputedStyle(canvas).getPropertyValue('width')
    , π = Math.PI;

  // Set some properties for the canvas to work with.
  this.ctx = canvas.getContext('2d');
  this.color = getComputedStyle(atom).getPropertyValue('color');
  this.text = atom.getElementsByTagName('input')[0];
  this.value = 0;
  this.parent = parent || atom;

  // Some defaults required for the animations
  this.r = +size.replace('px', '') / 2;
  this.ψ = π / 50;

  // Set height and width on the canvas.
  canvas.setAttribute('width', size);
  canvas.setAttribute('height', size);

  // Initialize and start the animation cycle for the first time.
  this.initialize(atom).animate(this.value, +this.text.value);
}

/**
 * Initialize event listeners and add input elements.
 *
 * @param {Element} atom
 * @returns {Atomic} fluent interface
 * @api private
 */
Atomic.prototype.initialize = function initialize(atom) {
  var electrons = atom.getElementsByTagName('li')
    , attributes = { type: 'radio', name: 'atomic' }
    , element, step, checked, n, i;

  // Attach onChange listeners to the text element
  this.text.addEventListener('change', this.listen.bind(this));

  // Step size and radio buttons.
  this.i = i = electrons.length;
  this.n = n = 100 / i;
  this.radio = [];

  // Set data-n on section to #list items.
  atom.setAttribute('data-n', i);

  // Add radiobuttons and listen to any clicks on the labels
  while (i--) {
    this.radio.unshift(element = document.createElement('input'));
    element.addEventListener('click', this.listen.bind(this));

    step = n * i;
    checked = step - this.text.value;

    attributes.id = 'i' + i;
    attributes.value = step;
    element.checked = checked > 0 && checked <= n && +this.text.value;

    this.parent.insertBefore(this.setAttributes(element, attributes), this.parent.firstChild);
    electrons[i].getElementsByTagName('label')[0].setAttribute('for', 'i' + i);
  }

  return this;
};

/**
 * Small helper function to set attributes on created elements.
 *
 * @param {Element} element
 * @param {Object} attributes
 * @returns {Element}
 * @api private
 */
Atomic.prototype.setAttributes = function setAttributes(element, attributes) {
  for(var key in attributes) {
    element.setAttribute(key, attributes[key]);
  }

  return element;
};

/**
 * Initialize animate function with specified start/end.
 *
 * @param {Number} start
 * @param {Number} end
 * @returns {Element}
 * @api private
 */
Atomic.prototype.animate = function animate(start, end) {
  var self = this
    , cw = end > start ? 1 : -1
    , π = Math.PI
    , Δ = this.ψ * cw
    , θ = Δ / 2
    , α = π / -2 + this.ψ * start
    , β = π / -2 + this.ψ * end
    , steps = Math.abs(Math.ceil((β - α) / Δ))
    , ctx = this.ctx
    , r = this.r
    , λ = 0;

  /**
   * Render each slice.
   *
   * @api private
   */
  function render() {
    var corr = (start && start !== 100) || λ ? θ : 0
      , r1 = α + λ * Δ - corr
      , r2 = r1 + corr + Δ;

    // Limit animation, final step might be one to much.
    if (~cw && r2 > β) r2 = β;
    if (!~cw && r2 < β) r2 = β;

    // Start drawing the circle slice.
    ctx.beginPath();
    ctx.globalCompositeOperation = 'destination-' + (~cw ? 'over' : 'out');
    ctx.fillStyle = ~cw ? self.color : 'rgba(0, 0, 0, 1)';
    ctx.arc(r, r, r, r1, r2, !~cw);
    ctx.lineTo(r, r);
    ctx.closePath();
    ctx.fill();

    // Update the current value to reflect the running endpoint of the animation.
    self.value = start + 1 * cw + λ++ * cw;

    // stop calling render as soon as steps are done or if equal to end.
    if (λ < steps && r2 !== β) return self.id = window.requestAnimationFrame(render);

    // If complet animation is rendered make sure the value is the provided end.
    // This will fix some rendering artifacts due to rounding errors.
    self.value = end;
  }

  this.id = window.requestAnimationFrame(render);
};

/**
 * Change the current value of the progress bar and selected radio buttons.
 *
 * @param {Number} end
 * @returns {Atomic} fluent interface
 * @api public
 */
Atomic.prototype.update = function update(end) {
  if (end === this.value) return this;
  if (end > 100) end = 100;

  // Which radio button should be selected.
  var current = Math.floor(end / this.n)
    , max = this.i - 1;

  // Update the input value and selected radio button.
  end = Math.round(end);
  this.text.value = end;
  this.radio[current < max ? current : max].checked = true;

  // Stop running animations and start fresh one.
  window.cancelAnimationFrame(this.id);
  this.animate(this.value, end);

  return this;
};

/**
 * Event listener for any changes on the inputs.
 *
 * @param {Event} e
 * @returns {Atomic} fluent interface
 * @api private
 */
Atomic.prototype.listen = function listen(e) {
  this.update(+(e.target || e.srcElement).value);

  return this;
};

/**
 * Initialize each progress bar on the page.
 *
 * @api public
 */
Atomic.init = function init() {
  var atoms = document.querySelectorAll('[data-atomic]');

  for (var k = 0; k < atoms.length; k++) {
    Atomic[k] = new Atomic(atoms[k]);
  }
};

//
// Expose module.
//
module.exports = Atomic;
},{}]},{},[1]);
